### PR TITLE
Visual regression test with permissions

### PR DESCRIPTION
**Pull request by jsoref**
I suspect this is the explanation for https://github.com/openameba/spindle/pull/645#issuecomment-1396540064

Note that this PR should be considered very scary. What you had previously was dangerous in that it was a workflow with full write permissions that could run from a PR (like this one), but because of the logic errors being fixed here, it didn't do everything that it was trying to do (namely conditionally run for PRs from forks).

With the changes here, you will be *running* the untrustworthy code (or in places the merge of the untrustworthy code).

Thus, if you are going to do this, you should make sure that:
1. `secrets.GOOGLE_APPLICATION_CREDENTIALS_JSON` can *only* be used to write to `secrets.GCS_BUCKET_NAME_REG_SUIT` (probably without being able to delete existing content)
2. `secrets.REG_SUIT_NOTIFY_CLIENT_ID` is a relatively harmless thing for someone to have access to

Also note that you can't test this PR from this PR (the event for a `pull_request_target` is dispatched to the version of the workflow in the target branch -- it is not sent to the version of the file as seen in the head of the pr or the merge of the PR), if you're interested in testing it, you should make another repository with this code and a second repository that's a fork of that repository and create a PR from the second repository to the first.

Note that to the greatest extent possible, it's much better to have a distinct job w/ elevated privileges that's responsible for processing the output generated from an untrusted code run as opposed to giving untrusted code access to secrets/write permissions. -- As you're already using `needs`, adding an extra `needs` isn't a big deal -- you can use `actions/upload-artifact` and `actions/download-artifact` or similar to pass data between the untrustworthy job and the writing job.

### CODE REVIEW

**Comment by herablog** [comment]
Thank you for your thoughtful suggestion. The word "with full write permissions" in your comment, are you referring to GITHUB_TOKEN permissions?

We have checked the points you mentioned and they seem to be fine, so we plan to merge it later. Also, we will consider using the `actions/upload-artifact` you suggested, as it seems to be better for this repository.

@sasaplus1 Can you double check this?

**Comment by jsoref** [comment]
Yes, I'm referring to the `GITHUB_TOKEN`

File: .github/workflows/visual-regression-test.yml [hunk 1]
```diff
@@ -77,7 +77,6 @@ jobs:
           --outDir packages/spindle-ui/.reg-suit/actual_images/
           --serverCmd 'npx serve -p 5000 packages/spindle-ui/public'
           http://localhost:5000
-      - run: git checkout ${{ github.head_ref || github.ref_name }}
```
Comment (sasaplus1) [inline .github/workflows/visual-regression-test.yml:80]: This line is probably necessary for reg-suit:
https://github.com/reg-viz/reg-suit#workaround-for-detached-head

Comment (jsoref) [reply .github/workflows/visual-regression-test.yml:80]: Ah. Let's give the step a name so people understand its importance.

**Review by sasaplus1 — APPROVED** [review]
Thank you for fixing!

**Review by herablog — APPROVED** [review]
The changes in this PR look fine. appreciate it.

**Pull request merged**